src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+            type Output = $ty;
+
+            #[inline]
+            fn div(self, rhs: $int) -> Self {
+                let rhs = Uint256::from(u64::from(rhs));
+                $ty(self.0 / rhs)
+            }
+        }
+    };
+}
+
+/// Implements `ops::Shr` and `ops::Shl` by integer type `$ty`. Silently casts shift argument to
+/// `usize`, this is safe as long as `$ty` never has more bits than the underlying architecture.
+macro_rules! impl_shift {
+    ($ty:ident, $int:ident) => {
+        impl Shl<$int> for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn shl(self, shift: $int) -> Self { Self(self.0 << shift as usize) }
+        }
+
+        impl Shr<$int> for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn shr(self, shift: $int) -> Self { Self(self.0 >> shift as usize) }
+        }
+    };
+}
+
+/// A little-endian 256 bit integer representing work.
+///
+/// Work is a measure of how difficult it is to find a hash below a given [`Target`].
+///
+/// ref: <https://en.bitcoin.it/wiki/Work>
+///
+/// # Examples
+///
+/// ```
+/// use bitcoin::Work;
+/// let a = Work::from(0xdeadbeef);
+/// let b = Work::from_le([0xdeadbeef, 0x00, 0x00, 0x00]);
+/// assert_eq!(a, b)
+/// ```
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct Work(Uint256);
+
+impl Work {
+    /// Lowest possible work value for Mainnet. See comment on [`Params::pow_limit`] for more info.
+    pub const MAINNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0000_ffff_0000_u64]));
+
+    /// Lowest possible work value for Testnet. See comment on [`Params::pow_limit`] for more info.
+    pub const TESTNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0000_ffff_0000_u64]));
+
+    /// Lowest possible work value for Signet. See comment on [`Params::pow_limit`] for more info.
+    pub const SIGNET_MIN: Work = Work(Uint256([0, 0, 0, 0x0000_0377_ae00_0000_u64]));
+
+    /// Lowest possible work value for Regtest. See comment on [`Params::pow_limit`] for more info.
+    pub const REGTEST_MIN: Work = Work(Uint256([0, 0, 0, 0x7fff_ff00_0000_0000_u64]));
+
+    /// Converts this [`Work`] to [`Target`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Work;
+    /// let w = Work::from(0xdeadbeef);
+    /// assert_eq!(w.to_target().to_work(), w)
+    /// ```
+    pub fn to_target(&self) -> Target { Target(self.0.inverse()) }
+}
+
+impl Add for Work {
+    type Output = Work;
+    fn add(self, rhs: Self) -> Self { Work(self.0.wrapping_add(rhs.0)) }
+}
+
+impl Sub for Work {
+    type Output = Work;
+    fn sub(self, rhs: Self) -> Self { Work(self.0.wrapping_sub(rhs.0)) }
+}
+
+do_impl!(Work);
+
+impl_mul!(Work, u8);
+impl_mul!(Work, u16);
+impl_mul!(Work, u32);
+
+impl_div!(Work, u8);
+impl_div!(Work, u16);
+impl_div!(Work, u32);
```

Comment:
Looks like you could've saved duplication by putting mul and div into same macro and calling `impl_mul_div` from `do_impl`

Reply:
At other iterations I' was supporting different sized integers for each, I'm still not totally sure I have all the correct (sane/meaningful) integers supported. Can merge them though if `u8` through `u32` is all we want for both.

Reply:
We now have `impl_div_mul` for ints sized 8 to 32

Reply:
And now also `u64`.